### DISCUSSION TITLE
Support M2A Corresponding Fields

### DISCUSSION THREAD

**Discussion by julbd** [directus/directus#9350] [Feature Requests]
Add support for creating "Corresponding fields" with Many-to-Any fields, like it is currently possible with Many-to-Many fields.
- A corresponding field form is displayed for each related collection selected.
- The corresponding field would be a **Any-to-Many** relation, with the `collection` field fixed to its collection name.

> **Example** : In a `Shelf` collection, the `items` field can contains many things (books, boxes, etc...). Some or all of these collections has a corresponding `shelves` field that display the shelves the book/box/etc are put in.
> Currently, with a "Many-to-Many" relation in corresponding collections, all the shelves corresponding to the many books/boxes/etc with the same id would be found.

<img width="634" alt="Screenshot 2021-11-02 at 14 15 59" src="https://user-images.githubusercontent.com/64699408/139854661-f9d466d8-d62b-4ea8-a95c-022932e1da80.png">

 
 
**Alternative solution** (simpler but error-prone) :
- Allow to manually set `collection`of the Many-to-Any field creation form, allowing to target the "Many"-side collections instead of the "Any"-side ones.

**Comment by rijkvanzanten** [comment]
Before adding the corresponding fields boolean flag, we'll have to support actual "any-to-many" relationship types first!

Also wondering if it would be "any to many", "one to any", "any to one" or.. (i'm getting my brain in a knot)

**Reply by julbd** [reply]
Updated my proposal :)

**Reply by julbd** [reply]
It should be "Any-to-Many" as it has to support junction collections with `id`+`collection`+`item` fields (with `collection` fixed to its own name). "One-to" or "-to-One" types have no junctions.

**Reply by JonL1** [reply]
> Also wondering if it would be "any to many", "one to any", "any to one" or.. (i'm getting my brain in a knot)

Any to none would be fun to implement 😆

**Comment by Dominic-Marcelino** [comment]
Did anyone found a solution for this?
Have a m2a relation and need to display the usage on the many side.

Example: 
`page` m2a to `block1` and `block2`, --> `blocks[1|2]` should have a list in which `page` they are used

**Reply by rijkvanzanten** [reply]
The "reverse" of an `a2o` type (eg `o2a`) doesn't exist yet in the API, so there's currently no way to get that information

**Reply by Dominic-Marcelino** [reply]
> The "reverse" of an `a2o` type (eg `o2a`) doesn't exist yet in the API, so there's currently no way to get that information

Thanks for clarification! Is this something that is generally planned for some time the future or way to specific? (Without giving any timeframe)

To make sure I’m not thinking into a completly wrong direction:
Would alternatively speak anything against creating an abstracted m2m interface, that uses the own collection in combination with the id to get those items?

**Reply by rijkvanzanten** [reply]
> Thanks for clarification! Is this something that is generally planned for some time the future or way to specific? (Without giving any timeframe)

It is, though it doesn't really come up all that often, so it's been a little tricky to prioritize. We are currently reworking a lot of the db abstraction logic though, so I'm hopeful we can "unlock" this in that process as well. 

> Would alternatively speak anything against creating an abstracted m2m interface, that uses the own collection in combination with the id to get those items?

That could be possible, but you'd lose the "m2a"-ness of the relationship as it would be "locked" to one of the related items right? 🤔

**Reply by Dominic-Marcelino** [reply]
> It is, though it doesn't really come up all that often, so it's been a little tricky to prioritize. We are currently reworking a lot of the db abstraction logic though, so I'm hopeful we can "unlock" this in that process as well.

Great news, looking forward to all upcoming functions! 


> 
> That could be possible, but you'd lose the "m2a"-ness of the relationship as it would be "locked" to one of the related items right? 🤔

Thanks for the input! That's true, but in our case the m2a would be used as a o2a and therewith a reversed a2o would be needed (from a single collection it would be a o2o)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by emahuni** [comment]
I was surprised that this doesn't exist. All the information to implement this is there as far as I can see. The logic is supposed to join using something like: 

```sql
... WHERE M2AJunctionTable.collection = 'TheAnyTable' AND M2AJunctionTable.item = TheAnySidePriKey
```

That way we have resolution to the A2M item.

I don't know how the whole DB rework out is looking, but this is the basic solution to it in its current form.

On another note, maybe this can require a separate discussion, but since you mentioned DB rework, let me just say something here:
I also think there should be a `O2O`, `hasMany` and `hasOne` relationships dedicated in the relationships selection, which allows clear and direct usage without figuring it out from abstract rules. Remember, since this is working on existing DBs, some use a field with an array of primary keys to do a `hasMany` relationship; eg: `phones: [4,23,8]` means that the field has many phones in this case ones with IDs `4, 23 and 8`. No other tables required. Plus the `hasOne` is just the simple one of this; `phone: 34`, just like the `M2O`, is expected to be just like that, `hasOne`, no other relations or tables required. The `hasOne` can be named `N2O` and `hasMany` can be named `N2M`, with `N` standing for `None` as per naming convention.

**Comment by emahuni** [comment]
I have managed to do it manually with the current implementation using an M2M. 

So I have a collection called `phonenumbers` that has an M2A field called `owners`. I want a collection called `banks_branches` to have a A2M `phones` field that refers back to the phone number that has `banks_branches` as the owner. To create that I have created the following SQL queries that create the `A2M` relationship and appropriate field in `banks_branches`.

```sql
INSERT INTO "public"."directus_fields" ("collection", "field", "special", "interface", "options", "display", "display_options", "readonly", "hidden", "sort", "width", "translations", "note", "conditions", "required", "group", "validation", "validation_message") VALUES ('banks_branches', 'phones', 'm2m', 'list-m2m', NULL, NULL, NULL, 'f', 'f', NULL, 'half', NULL, NULL, NULL, 'f', NULL, NULL, NULL);

INSERT INTO "public"."directus_relations" ( "many_collection", "many_field", "one_collection", "one_field", "one_collection_field", "one_allowed_collections", "junction_field", "sort_field", "one_deselect_action") VALUES ( 'phonenumbers_owners', 'item', 'banks_branches', 'phones', NULL, NULL, 'phonenumbers_id', NULL, 'nullify');
```

Running this query inserts those rows and everything works as if this is meant to work like this. You can create phones and bank branches from any side and manage these just like `M2M` relationships. Below is the `phonenumbers` `M2A` side of the relationship:

![image](https://github.com/directus/directus/assets/7290508/bfafa796-5640-43b3-8671-e7e61177b5d2)

Below is the `banks_branches` `A2M` side of the relationship:

![image](https://github.com/directus/directus/assets/7290508/3577796e-1c39-4f44-bbbf-c8ef0e075a6b)

The only problem here is selecting the correct collection in `phonenumbers_owners` as this will display even phonenumbers from other `owners`, still trying to figure out how to use filters to do that or some other way to do it.  Above, the `phonenumber` with `ID` `2` is not owned by a `banks_branches` collection, this is just to demonstrate what I mean.

I tried to change the first line to the following to solve that and try again, but it somehow causes issues and detects that the relationship is fishy and hacky.

```sql
INSERT INTO "public"."directus_fields" ("collection", "field", "special", "interface", "options", "display", "display_options", "readonly", "hidden", "sort", "width", "translations", "note", "conditions", "required", "group", "validation", "validation_message") VALUES ('banks_branches', 'phones', 'm2m', 'list-m2m', '{"filter":{"_and":[{"owners":{"collection":{"_eq":"banks_branches"}}}]}}', NULL, NULL, 'f', 'f', NULL, 'half', NULL, NULL, NULL, 'f', NULL, NULL, NULL);
```

Note the filter: `{"filter":{"_and":[{"owners":{"collection":{"_eq":"banks_branches"}}}]}}`

Please if you know how to let me know. This is my first rodeo with Directus. I will of course update if I manage to do it.

**Reply by emahuni** [reply]
I know I am kinda late to the party, but please hear me out; this is needed it complements the relationship properly.
